archive/src/test.rs

```diff
@@ -1,4 +1,80 @@
 //! Unit tests.
 
+use std::{thread::sleep, time::Duration};
+
+use crate::{expand::walk, Target, DEFAULT_ARCHIVE_POSTFIX};
+
 #[test]
-fn feature() {}
+fn walk_removes_dirs() {
+    let target = Target::builder().root("testdata/simplezip").build();
+    assert!(target.root().exists());
+
+    let walker = walk(target, Default::default());
+
+    let mut temp_paths = Vec::new();
+    for entry in walker {
+        let entry = entry.expect("walk entry");
+        let is_archive_child = entry
+            .path()
+            .to_string_lossy()
+            .contains(DEFAULT_ARCHIVE_POSTFIX);
+        if is_archive_child {
+            // Instead of pushing `Entry`, push its constituent paths and drop the entry.
+            // This is required because `Entry` keeps its containing directory alive until all entries have been dropped.
+            temp_paths.push((entry.concrete().to_owned(), entry.path().to_owned()));
+        }
+    }
+
+    // Let some time pass: temp directories are cleaned up on drop, so this process may not be instant.
+    sleep(Duration::from_millis(100));
```

Comment:
[nit]: Rust guarantees that if the application does not panic, the destructor will be run (since there's no background thread to run GC on).  Rust requires that we do not rely on destructors running to guarantee soundness, but we can use it to enforce logic.  We should be able to guarantee that cleanup occurs by this point by inserting a `drop` here (unless the value is consumed earlier).  I just want to avoid relying on time during tests.

If that is still flaky, I'd prefer to run an idempotent manual cleanup (using the same method that `Drop` uses) to confirm that we've cleaned up correctly.

Reply:
You know, I was implicitly modeling `Drop` as running "sometime in the background after the item is dropped" but actually you're right, it runs in the foreground: https://abramov.io/rust-dropping-things-in-another-thread

I'll just get rid of these delays, I doubt they're needed based on that.

Reply:
I missed it at first too, until I started thinking about ways to avoid this sleep.  We've been poisoned by GC languages.

Reply:
FYI CPython also does this, since they exclusively use reference-counted GC and drop unreferenced values immediately after the last ref is dropped.